### ISSUE TITLE
Do not allow surrounding spaces to be counted as part of ${number} templates in template index strings 

### ISSUE THREAD

**Issue by craigphicks** [microsoft/TypeScript#46109]
# Suggestion

## 🔍 Search Terms

Template String Pattern Index Signatures
Template Index Strings

## ✅ Viability Checklist

My suggestion meets these guidelines:

* [ ] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

Actually it would change the typescript behavior as it was released - maybe better now than later?
The current behavior might be more bug than feature.

## ⭐ Suggestion

Not allowing spaces would reduce typing mistakes.

## 📃 Motivating Example

Modification of the [example feature explanation in  the handbook](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#symbol-template-signatures):

```typescript
interface Options {
    width?: number;
    height?: number;
}

let a: Options = {
    width: 100,
    height: 100,
    "data-1": true, // Error! 'data-blah' wasn't declared in 'Options'.
};

interface OptionsWithDataProps extends Options {
    // Permit any property matching 'data-${number}'.
    [optName: `data-${number}`]: unknown;
}

let b: OptionsWithDataProps = {
    width: 100,
    height: 100,
    "data-1": true,       // Works!
    "data- 1": true,       // Passes but might be better off failing
    "data-1 ": true,       // Passes but might be better off failing
    "data- 1 ": true,       // Passes but might be better off failing
    "data- ": true,       // Passes but might be better off failing  (no number at all)

    "unknown-property": true,  // Error! 'unknown-property' wasn't declared in 'OptionsWithDataProps'.
};
```
[playground](https://www.typescriptlang.org/play?#code/FDCWDsBcFMCcDMCGBjaACA8gB0qA9uAM5oDewaFaA7qACaQAWA-AFxrgCuAtgEZwDc5Sg2igA5g0it23PrEEBfEABtokNIjbZcBYgF5SQijXoM2ARgAMlgDRG0I8ZIvW7lNACJaiSIgC05h5skLAc0DZoAPSRaACisLB4sACEaADk3r5+PMqIDGnUiITgaeq00Mi5sNC0aBDp2vhEaQB0wAqCYFBwSKiYOE2EAOqgjAAiPogAColYxNAAHjDgtMSNuobu0WhTcFyjGuAAnmhYs3CQJ1w+yAwQYumZ-gAkJJy8cAqt9gDaeDgAOUQXGgbAABk8-K93nIFGCALpsDjgADW4DwVHAihUajQPC0A10I3Gkxm-30m0oJkYLls9kcEkgtLclC8kwCQTQITCEXcFG2QySKMIyXsbKyaECwVC4T5-JiUyKhGgxB4HHU+0ZePQfEgMFgaDw8HgaCQoGU9zFkPMnmlPLlUQVSpVePVaE1km12r1cENxtNiHNlvc4v8kttXJlvL520VhGVqrdHvUfG9+r9JrNFvAYit7Ij3NlctjzsTGqcKZ1anTRszgezubFyLRGPAfjO-wuR05hd523iiRS6Wb6Mx7fOsEuBSoRRKZQqVRqdXADUJRGJDAmvjJc2+HRAQA)

## 💻 Use Cases

Using it to catch unintended key entries.

**Comment by IllusionMH** [comment]
Can be simplified to
```ts
const a: `data-${number}` = "data-1"; // OK

// No errors
const a1: `data-${number}` = "data- 1";
const a2: `data-${number}` = "data-1 ";
const a3: `data-${number}` = "data- 1 ";
const a4: `data-${number}` = "data- ";
```

**Comment by jcalz** [comment]
Oh wow this an even better example for my [list of surprising behaviors of `` `${number}` ``](https://github.com/microsoft/TypeScript/issues/41893#issuecomment-889292665)!  I'm still not quite able to internalize the interpretation of `` `${number}` `` meaning "some string that [converts](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-runtime-semantics-stringnumericvalue) to a finite number",  but apparently that's what it means (as opposed to "the type of `` `${n}` `` when `n` is a finite number).

**Comment by craigphicks** [comment]
@jcalz - Thanks for link to [7.1.4.1.2 Runtime Semantics: StringNumericValue](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-runtime-semantics-stringnumericvalue), that is very informative.  So the behavior is intended to match some formal standard - as I suspected but wasn't sure.  (I also enjoyed you list of examples!) 

For a user on the consuming end of the data, if they want to correctly decode the pattern suffix to disambiguate the many-to-one mapping of strings to numbers, they must use of algorithm listed in `[7.1.4.1.2 Runtime Semantics: StringNumericValue]`.  I see that a possible source of confusion and errors.  That risk would be mitigated by employing a simple one-to-one mapping.

Having a many-to-one mapping of keys strings to semantic keys, before using the keys to lookup the value in the object map seems anti intuitive and error prone.  But the original problem lies in JS, not TS.

On the other hand, if TS is going to impose some strict typing on the keys for arbitrary integers, why not go the whole way and allow specification of a one-to-one map?   For an explicit limited range of integers it is already possible:

```
type CountingIntegers = "1"|"2";
type FooN = Record<`foo${CountingIntegers}`,any>
```

Maybe the feature request should be for a temple-key-string use specific intrinsic utility type "CountingIntegers" allowing all nonnegative integers up to `Number.MAX_SAFE_INTEGER`.  That is narrow enough to be safe and easy to implement, is extremely simple for a user to understand completely at a glance, and has practical value in eliminating misunderstandings.

**Comment by craigphicks** [comment]
Thanks to @jcalz pointing out  [7.1.4.1.2 Runtime Semantics: StringNumericValue](https://tc39.es/ecma262/multipage/abstract-operations.html#sec-runtime-semantics-stringnumericvalue), it is now clear within a shadow of a doubt that the "surrounding space" behavior is "working as intended".  So I will close this issue.

**Comment by jcalz** [comment]
I’m not sure… I’d really like to hear someone officially say that this is intentional.  I’m just an interested bystander.

**Comment by IllusionMH** [comment]
Same here regarding whitespace.
While `100.1` or even `1e3` are really looking as number literals including whitespace doesn't look expected from usage perspective, but that's IMHO.

**Comment by andrewbranch** [comment]
This is very surprising to me and seems undesirable, but I’ll let @ahejlsberg have the definitive word.

**Comment by ahejlsberg** [comment]
Yeah, we're basically just relying on JavaScript's intrinsic string-to-number conversion to validate `` `${number}` `` placeholder values--and it permits leading and trailing whitespace. Also, strangely, it permits empty strings and strings with only whitespace and returns `0` for those.

I agree it makes more sense to not permit any whitespace and in particular to not permit empty strings. It's technically a breaking change, but can't imagine many types depend on it.

**Comment by nnmrts** [comment]
> Yeah, we're basically just relying on JavaScript's intrinsic string-to-number conversion to validate `` `${number}` `` placeholder values--and it permits leading and trailing whitespace. Also, strangely, it permits empty strings and strings with only whitespace and returns `0` for those.
> 
> I agree it makes more sense to not permit any whitespace and in particular to not permit empty strings. It's technically a breaking change, but can't imagine many types depend on it.

@ahejlsberg 

Okay, but why is TypeScript then handling empty strings and whitespace strings differently?

```ts
type ThisIsTrue = " " extends `${number}` ? true : false;
type ThisIsFalse = "" extends `${number}` ? true : false;

const thisIsZero = Number(" ");
console.log(thisIsZero);

const thisIsAlsoZero = Number("");
console.log(thisIsAlsoZero);
```

https://www.typescriptlang.org/play/?#code/C4TwDgpgBAKgFgSwM4EkkwE4FdoF4oBEhUEAHsBAHYAmSUABgCQDelWAtgEYQYC+9UAPxRg2aAC4oAMwCGAGyQQA3AChQkWIlRIAYvMVR8BImQo06TVh258Bw0TiiTZC5SpUBjAPaUkwEVpoAFo8XoZQAHLWPAAURAQAlKrevl5yEAB0cl4A5jHAgUghGF5J7il+AchoAIIKXsVh+FFcscZlFWmZ2XkF1Uh1SA2hSUA
